rust/automerge/src/op_set2/op_set.rs

```diff
@@ -1177,24 +1177,72 @@ impl OpsFound<'_> {
         self.ops.last().map(|o| o.width(encoding)).unwrap_or(0)
     }
 
-    pub(crate) fn check_for_noop(&mut self, action: &mut types::OpType) -> bool {
+    /// Determine what action to take based on the found operations
+    ///
+    /// The action provided by the user may actually not be needed, or it may
+    /// not result in visible changes to the document. This method determines
+    /// what the `ResolvedAction` representing these cases should be and also
+    /// updates the `OpsFound::ops` where necessary.
+    ///
+    /// # Returns
+    ///
+    /// `Some(ResolvedAction)` if there is an op which needs to be inserted into
+    /// the opset, or `None` otherwise
+    pub(crate) fn resolve_action(
+        &mut self,
+        original_action: types::OpType,
+    ) -> Option<ResolvedAction> {
         if let Some(op) = self.ops.last() {
-            if let types::OpType::Put(v) = action {
+            if let types::OpType::Put(v) = &original_action {
```

Comment:
tiny nit - but I think you can `let Put(v) = ... {   } else { Some(ResolvedAction::VisibleUpdate(original_action))) }` and now all your if branches are covered and all the returns can go away.